### PR TITLE
8341408: Implement JEP 488: Primitive Types in Patterns, instanceof, and switch (Second Preview)

### PR DESCRIPTION
**Pull request by biboudis**
This PR prepares the improvement for JEP 488.

The proposed approach is described in the CSR.

In terms of compiler implementation, we add the related translation in `TransPatterns`. Interestingly, this issue would be ideally addressed in `TransTypes`. However, there are some conveniences in `TransPatterns` that would need to be pulled into `TransTypes` to make this work: 1) the dependency to `makeBinary` and transitively to `operators` and 2) a lot of ceremony around `currentMethodSym` is already setup nicely in `TransPatterns`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a JEP request to be targeted
- [x] Change requires CSR request [JDK-8342397](https://bugs.openjdk.org/browse/JDK-8342397) to be approved

### Issues
 * [JDK-8341408](https://bugs.openjdk.org/browse/JDK-8341408): Implement JEP 488: Primitive Types in Patterns, instanceof, and switch (Second Preview) (**Sub-task** - P4)
 * [JDK-8335876](https://bugs.openjdk.org/browse/JDK-8335876): JEP 488: Primitive Types in Patterns, instanceof, and switch (Second Preview) (**JEP**)
 * [JDK-8342397](https://bugs.openjdk.org/browse/JDK-8342397): Implement JEP 488: Primitive Types in Patterns, instanceof, and switch (Second Preview) (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21539/head:pull/21539` \
`$ git checkout pull/21539`

Update a local copy of the PR: \
`$ git checkout pull/21539` \
`$ git pull https://git.openjdk.org/jdk.git pull/21539/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21539`

View PR using the GUI difftool: \
`$ git pr show -t 21539`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21539.diff">https://git.openjdk.org/jdk/pull/21539.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21539#issuecomment-2427031548)
</details>

### CODE REVIEW

File: test/langtools/tools/javac/patterns/PrimitiveTypesInTestingContextErasure.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import java.util.List;
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/patterns/PrimitiveTypesInTestingContextErasure.java:1]: nit: this import seems to be out of place

File: test/langtools/tools/javac/patterns/PrimitiveTypesInTestingContextErasure.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+/*
+ * @test
+ * @bug 8341408
+ * @summary Compiler Implementation for Primitive types in patterns, instanceof, and switch (Second Preview)
+ * @enablePreview
+ * @compile PrimitiveTypesInTestingContextErasure.java
```
[the remark this answers is not in the document]
Comment (biboudis) [reply test/langtools/tools/javac/patterns/PrimitiveTypesInTestingContextErasure.java]: Done

File: test/langtools/tools/javac/patterns/PrimitiveTypesInTestingContextErasure.java (hunk 1 above)
Comment (biboudis) [reply test/langtools/tools/javac/patterns/PrimitiveTypesInTestingContextErasure.java]: Also moved it. Thx!

**Review by vicente-romero-oracle — APPROVED** [review]
looks sensible

**Review by lahodaj — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**